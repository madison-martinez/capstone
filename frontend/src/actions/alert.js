export const alertActionTypes = {
    SUCCESS: 'ALERT_SUCCESS',
    ERROR: 'ALERT_ERROR',
    RESET: 'ALERT_RESET'
};

function success(message) {
    return {
        type: alertActionTypes.SUCCESS,
        message
    }
};

function error(message) {
    return {
        type: alertActionTypes.ERROR,
        message
    }
};

function reset() {
    return {
        type: alertActionTypes.RESET
    }
};


export const alertActions = {
    success,
    error,
    reset
};

