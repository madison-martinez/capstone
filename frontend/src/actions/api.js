import axios from 'axios';

const baseUrl = "https://localhost:5001/api/";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
    product(url = baseUrl + 'Product/') {

        return {
            fetchAll: () => axios.get(url),
            create: newProduct => axios.post(url, newProduct),
            update: (id, updateProduct) => axios.put(url + id, updateProduct),
            delete: id => axios.delete(url + id),
            fetchById: id => axios.get(url + id)
        }
    }
};
