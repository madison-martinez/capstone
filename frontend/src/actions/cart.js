export const CartActionTypes = {
  ADD_TO_CART: "ADD_TO_CART",
  GET_CART_COUNT: "GET_CART_COUNT",
  DELETE_CART_PRODUCT: "DELETE_CART_PRODUCT",
  CLEAR_CART: "CLEAR_CART",
};

export const addToCart = ({
  product,
  id,
  price,
  description,
  quantity,
  image,
}) => {
  const cartProduct = { product, id, price, description, quantity, image };
  return (dispatch) => {
    dispatch({
      type: CartActionTypes.ADD_TO_CART,
      payload: cartProduct,
    });
  };
};

export const getCartCount = () => {
  return (dispatch) => {
    dispatch({
      type: CartActionTypes.GET_CART_COUNT,
    });
  };
};

export const deleteCartProduct = ({ id, price }) => {
  const deletedItem = { id, price };
  return (dispatch) => {
    dispatch({
      type: CartActionTypes.DELETE_CART_PRODUCT,
      payload: deletedItem,
    });
  };
};

export const clearCart = () => {
  return (dispatch) => {
    dispatch({
      type: CartActionTypes.CLEAR_CART,
    });
  };
};
