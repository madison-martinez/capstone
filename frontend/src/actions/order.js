
export const OrderActionTypes = {
  PLACE_ORDER: "PLACE_ORDER",
};

export const placeOrder = ({ products, price, userId }) => {
  const order = { products, price, userId };
  return async (dispatch) => {
    dispatch({
      type: OrderActionTypes.PLACE_ORDER,
      payload: order,
    });
  };
};
