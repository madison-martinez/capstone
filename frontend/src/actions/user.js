import { userService } from "../services/user.service";
import { alertActions } from "./alert";

export const userConstants = {
  REGISTER_REQUEST: "USERS_REGISTER_REQUEST",
  REGISTER_SUCCESS: "USERS_REGISTER_SUCCESS",
  REGISTER_FAILURE: "USERS_REGISTER_FAILURE",

  LOGIN_REQUEST: "USERS_LOGIN_REQUEST",
  LOGIN_SUCCESS: "USERS_LOGIN_SUCCESS",
  LOGIN_FAILURE: "USERS_LOGIN_FAILURE",

  LOGOUT: "USERS_LOGOUT",
};

function login(username, password, handleOnSuccess) {
  return async (dispatch) => {
    dispatch(request({ username }));

    return userService.login(username, password).then(
      (user) => {
        dispatch(success(user));
        handleOnSuccess();
      },
      (error) => {
        dispatch(failure(error.toString()));
        dispatch(alertActions.error(error.toString()));
      }
    );
  };

  function request(user) {
    return { type: userConstants.LOGIN_REQUEST, user };
  }
  function success(user) {
    return { type: userConstants.LOGIN_SUCCESS, user };
  }
  function failure(error) {
    return { type: userConstants.LOGIN_FAILURE, error };
  }
}

function logout() {
  userService.logout();
  return { type: userConstants.LOGOUT };
}

function register(user, handleOnSuccess) {
  return async (dispatch) => {
    dispatch(request(user));

    userService.register(user).then(
      (user) => {
        dispatch(success(user));
        handleOnSuccess();
      },
      (error) => {
        dispatch(failure(error.toString()));
      }
    );
  };

  function request(user) {
    return { type: userConstants.REGISTER_REQUEST, user };
  }
  function success(user) {
    return { type: userConstants.REGISTER_SUCCESS, user };
  }
  function failure(error) {
    return { type: userConstants.REGISTER_FAILURE, error };
  }
}

export const userActions = {
  login,
  logout,
  register,
};
