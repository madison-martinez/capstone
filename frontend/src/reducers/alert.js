import { alertActionTypes } from "../actions/alert";

export function alert(state = {}, action) {
  switch (action.type) {
    case alertActionTypes.SUCCESS:
      return {
        type: "ALERT_SUCCESS",
        message: action.message,
      };
    case alertActionTypes.ERROR:
      return {
        type: "ALERT_ERROR",
        message: action.message,
      };
    case alertActionTypes.RESET:
      return {};
    default:
      return state;
  }
};
