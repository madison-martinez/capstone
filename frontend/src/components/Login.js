import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { useRouter } from 'next/router';
import useForm from '../utils/useForm';
import Form from './styles/FormStyles';
import { userActions } from '../actions/user';


function LoginPage() {
    const { values, handleChange, clearForm } = useForm({
        username: '',
        password: '',
    });
    const [submitted, setSubmitted] = useState(false);
    const { username, password } = values;
    const dispatch = useDispatch();
    const router = useRouter();

    useEffect(() => {
        dispatch(userActions.logout());
        router.prefetch('/marketplace');
    }, []);

    return (
        <>
        <Form
            method="post"
            onSubmit={async e => {
                e.preventDefault();
                setSubmitted(true);
                if (username && password) {
                    dispatch(userActions.login(username, password));
                    router.push('/marketplace');
                }
                clearForm();
            }}
        >
            <fieldset>
                <h2>Login to your account</h2>
                <label htmlFor="username">
                    Username
                <input
                        type="username"
                        name="username"
                        placeholder="Username"
                        value={values.username}
                        onChange={handleChange}
                    />
                    {submitted && !username &&
                        <div className="invalid-feedback">Username is required</div>
                    }
                </label>
                <label htmlFor="password">
                    Password
                 <input
                        type="password"
                        name="password"
                        placeholder="Password"
                        value={values.password}
                        onChange={handleChange}
                    />
                    {submitted && !username &&
                        <div className="invalid-feedback">Password is required</div>
                    }
                </label>
                <button type="submit">Sign In!</button>
            </fieldset>
        </Form>
        </>
    );
};

export default LoginPage;