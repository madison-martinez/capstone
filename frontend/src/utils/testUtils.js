import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { reducers } from '../reducers';

function render(
  ui,
  {
    initialState,
    store = createStore(reducers, initialState ),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

//mock middleware functions
const thunk = ({ dispatch, getState }) => next => action => {
    if (typeof action === 'function') {
      return action(dispatch, getState)
    }
    return next(action)
};

const fakeItem = () => ({
    id: '1',
    price: 20.00,
    title: 'Roses',
    description: 'Roses are Red',
});

export {
    fakeItem,
    thunk
};

export * from '@testing-library/react';
export { render };