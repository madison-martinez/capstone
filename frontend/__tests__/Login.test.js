import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import { rest } from "msw";
import { setupServer } from "msw/node";
import fetchMock from 'fetch-mock';
import userEvent from "@testing-library/user-event";
import { waitFor } from "@testing-library/react";
import { useRouter } from 'next/router';
import Login from "../src/components/Login";
import { fakeUser, render, screen } from "../src/utils/testUtils";
import { alertActionTypes } from "../src/actions/alert";
import { userConstants, userActions } from "../src/actions/user";
import { userService } from "../src/services/user.service";

const middleware = [thunk];
const mockStore = configureStore(middleware);
const initialState = {
  authentication: {
    user: {},
  },
};

const mockPush = jest.fn();
jest.mock("next/router", () => ({
    useRouter: () => ({
        push: mockPush
    })
  }));

const mockSuccessFn = jest.fn();

const user = fakeUser();
//
const server = setupServer(
  rest.post("https://localhost:5001/users/authenticate", (req, res, ctx) => {
    return res(ctx.json(user));
  })
);

beforeAll(() => server.listen());
afterEach(() => {
  server.resetHandlers();
  jest.clearAllMocks();
  fetchMock.restore();
});
afterAll(() => server.close());

describe("<Login />", () => {
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it("should render empty form", () => {
    const { container } = render(<Login authUser={{ user: false }} />);
    expect(screen.queryByTestId("login-form")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  describe("inputting username and password", () => {
    it("handles user submitting a username and password", async () => {
      render(<Login store={store} />);
      expect(screen.queryByTestId("login-form")).toBeTruthy();

      userEvent.type(screen.getByPlaceholderText("Username"), user.username);
      userEvent.type(screen.getByPlaceholderText("Password"), user.password);

      userEvent.click(screen.getByText("Sign In!"));

      const expectedActions = [
        {
          type: userConstants.LOGIN_REQUEST,
          user: {username: user.username}
        }
      ];
      return store.dispatch(userActions.login(user.username, user.password, mockSuccessFn))
        .then(async () => {
          const actualActions = store.getActions();
          expect(actualActions).toEqual(expectedActions);
        })
    });

    it.only("reroutes to home page after successful login", () => {
        render(<Login store={store} />);
    
        expect(screen.queryByTestId("login-form")).toBeTruthy();
        userEvent.type(screen.getByPlaceholderText("Username"), user.username);
        userEvent.type(screen.getByPlaceholderText("Password"), user.password);
        userEvent.click(screen.getByText("Sign In!"));
        
        fetchMock.getOnce("/users/authenticate", {
            body: {user},
            headers: {"Content-Type": "application/json"}
        });

        const expectedActions = [
            {
                type: userConstants.LOGIN_REQUEST,
                user: {username: user.username}
              },
            {
                type: userConstants.LOGIN_SUCCESS,
                user
            }
        ];
        return store.dispatch(userActions.login(user.username, user.password, mockSuccessFn)).then(async() => {
            const actualActions = store.getActions();
            expect(actualActions).toEqual(expectedActions);
            await waitFor(() => expect(mockSuccessFn).toHaveBeenCalledTimes(1));
        })
    });
    
    it("shows error if username or password is incorrect", () => {});
  });
});
