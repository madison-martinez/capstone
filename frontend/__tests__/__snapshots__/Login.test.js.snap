// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render empty form 1`] = `
<div>
  <form
    class="FormStyles__Form-sc-1qtbzh2-0 cmkxQl"
    data-testid="login-form"
    method="post"
  >
    <fieldset>
      <h2>
        Login to your account
      </h2>
      <label
        for="username"
      >
        Username
        <input
          name="username"
          placeholder="Username"
          type="username"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          autocomplete="new-password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Sign In!
      </button>
    </fieldset>
    <img
      alt="fieldrows"
      height="500"
      src="/assets/photo-1464226184884-fa280b87c399.jpeg"
      width="500"
    />
  </form>
</div>
`;
